python/cuml/common/array_sparse.py

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from cuml.common.import_utils import has_scipy
+
+import cupyx as cpx
+import numpy as np
+
+from cuml.common.memory_utils import with_cupy_rmm
+
+from cuml.common.input_utils import input_to_cuml_array
+
+if has_scipy():
+    import scipy.sparse
+
+
+class SparseCumlArray:
+    """
+    SparseCumlArray abstracts sparse array allocations. This will
+    accept either a Scipy or Cupy sparse array and construct CumlArrays
+    out of the underlying index and data arrays. Currently, this class
+    only supports the CSR array format and input in any other sparse
+    format will be converted to CSR.
+
+    Parameters
+    ----------
+
+    data : scipy.sparse.spmatrix or cupyx.scipy.sparse.spmatrix
+        A Scipy or Cupy sparse matrix
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy or cupy data type.
+
+    Attributes
+    ----------
+
+    indptr : CumlArray
+        Compressed row index array
+    indices : CumlArray
+        Column indices array
+    data : CumlArray
+        Data array
+    dtype : dtype
+        Data type of data array
+    shape : tuple of ints
+        Shape of the array
+    nnz : int
+        Number of nonzeros in underlying arrays
+    has_sorted_indices : bool
+        Whether column indices and data are sorted by column
```

Comment:
How does has_sorted_indices come into play? Does it allow some optimizations, or... ?

Reply:
Vibhu brought this one to my attention. The KNN graph is a special case where the columns and values need to stay sorted by neighborhood distances. This is just a nice sanity check for us to use in our algorithms to make sure it has not been sorted previously. 

I need to think a little more about this one because I think setting it automatically could potentially cause some issues (like a knn_graph that just happens to be sorted naturally).